packages/bpk-component-infinite-scroll/src/BpkInfiniteScroll.js

```diff
@@ -0,0 +1,172 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import 'intersection-observer';
+
+import { BpkSpinner, SPINNER_TYPES } from 'bpk-component-spinner';
+
+import STYLES from './BpkInfiniteScroll.scss';
+
+const infinitize = ComponentToExtend => {
+  class Infinitize extends Component {
+    constructor(props) {
+      super(props);
+
+      this.state = {
+        index: 0,
+        elementsToRender: [],
+        isListFinished: false,
+        showSeeMore: false,
+      };
+
+      this.onRequestElements = props.onRequestElements;
+      this.onScroll = props.onScroll;
+      this.handleIntersection = this.handleIntersection.bind(this);
+      this.handleSeeMoreClick = this.handleSeeMoreClick.bind(this);
+      this.handleKeyPress = this.handleKeyPress.bind(this);
+      // TODO add polyfill for IE
+      this.observer = new IntersectionObserver(
+        (...args) => this.handleIntersection(...args),
+        { threshold: 1 },
+      );
+    }
+
+    componentDidMount() {
+      if (this.sentinel) {
+        this.observer.observe(this.sentinel);
+      }
+    }
+
+    componentDidUpdate() {
+      if (this.sentinel && this.state.index > 0) {
+        this.observer.observe(this.sentinel);
+      }
+    }
+
+    async loadMore() {
+      const { index, elementsToRender } = this.state;
+      const { elementsPerScroll, seeMoreAfter } = this.props;
+      const nextElements = await this.onRequestElements(
+        index,
+        elementsPerScroll,
+      );
+      if (nextElements && nextElements.length > 0) {
+        const nextIndex = index + elementsPerScroll;
+        return {
+          index: nextIndex,
+          elementsToRender: elementsToRender.concat(nextElements),
+          showSeeMore: seeMoreAfter === index / elementsPerScroll,
+        };
+      }
+      this.props.onScrollFinished({
+        totalNumberElements: elementsToRender.length,
+      });
+      return {
+        isListFinished: true,
+      };
+    }
+
+    handleIntersection(entries) {
+      entries.forEach(async entry => {
```

Comment:
do we really want `async` here? this means that the next iteration will only start after the current one is finished (including the async part). We can avoid this using the promise callback.

```
entries.forEach(entry => {
  if (entry.intersectionRatio === 1) {
    this.props.onScroll({ currentIndex: this.state.index });
    this.observer.unobserve(this.sentinel);
    this.loadMore().then(nextState => {
       this.setState(nextState);
     });
})
```

Also, as I mentioned above (in the tests) we need to return the promise here to make testing easier, something like:
```
return Promise.all(entries.map(...
```

Reply:
Actually, the `forEach` is not even necessary because there is only 1 entry (only one element been observed). 
Regarding returning the `Promise`, I'm testing the behaviour of the component, not the specific functions so I don't see the point of returning anything in the `handleIntersection` function.

Reply:
The point is so you can test the component without using setState like I mentioned above. If you call `setState` with `loadMore` you are not really testing how the component will be used or behaves in a real world situation. Furthermore, setState is an internal API, a good test should be unaware of the internal (private) implementation of then component. If in the future we decide to change the internal implementation but keep yielding the same result I expect the tests to pass without the need to change them too.

Reply:
That will be solved using an approach like the one you proposed above (calling `intersectionCallback`)

Reply:
Yes, but that only works if the promise is returned here so the test can wait for it before doing assertions. That's why I suggested you return it here 😃

Reply:
changed this function to be async (so would return a Promise that we can await in the tests). Also removed the `forEach` as we process only the first element.